File: python/topological_sort.py



graph = {
    2: [1, 3, 4],
    3: [1,4],
}

visited = {}
def topological_sort(graph, vertex,  arr = []):
    if not vertex:
        return
    if graph.get(vertex) is None and visited.get(vertex) is None:
        arr.append(vertex)
        visited[vertex] = True

    else:
        children = graph.get(vertex)
        if children is not None:
            for child in children:
                topological_sort(graph, child, arr)

        if visited.get(vertex) is None:
            arr.append(vertex)

    return arr

print(topological_sort(graph, 2, []))

File: C++/topological_sort.cpp

#include <algorithm>
#include "vector"

using namespace std;

typedef vector<vector<int>> vii;
typedef vector<int> vi;


void traverse(int v, vector<bool>& used, vii& g, vi& ans) {
    used[v] = true;
    for (size_t i = 0; i < g[v].size(); ++i) {
        int to = g[v][i];
        if (!used[to])
            traverse(to, used, g, ans);
    }
    ans.push_back(v);
}

void topological_sort(vii &graph, vector<bool> &visited, vi& ans) {
    for (int i = 0; i < graph.size(); ++i) {
        visited[i] = false;
    }
    ans.clear();
    for (int i = 0; i < graph.size(); ++i) {
        if (!visited[i]) {
            traverse(i, visited, graph, ans);
        }
    }
}

